### PR TITLE
feat:  PragmaDecoder + Hyperlane Test utils

### PR DESCRIPTION
**Pull request by JordyRo1**
## Implementation

 - Solidity contract implementation 
 - Tests 

## Flow details: 

 - On the test implemented, we first generate a feed_id full message (feed_id = asset_class + feed_type + pair_id stored in a bytes32). This message will contain the updateData for a single pair_id configuration. This message is fully decoded and if the validators signatures / merkle root matches , it is updated to the solidity contract storage
 
 ## Improvement 

 -  Currently the flow is working for a feed_id. We might need to add more tests afterwards

### CODE REVIEW

File: solidity/src/libraries/BytesLib.sol [hunk 1]
```diff
@@ -356,6 +356,17 @@ library BytesLib {
         return tempUint;
     }
 
+    function toInt256(bytes memory _bytes, uint256 _start) internal pure returns (int256) {
```
Comment (EvolveArt) [inline solidity/src/libraries/BytesLib.sol:359]: is that tested ? not sure of the implementation

File: solidity/src/libraries/DataParser.sol [hunk 2]
```diff
@@ -2,121 +2,46 @@
...
 library DataParser {
     using BytesLib for bytes;
 
-    uint16 constant SM = 21325;
-    uint16 constant TW = 21591;
-    uint16 constant RV = 21078;
-    uint16 constant OP = 20304; 
-    uint16 constant PP = 20560;
-
     function parse(bytes memory data) internal pure returns (ParsedData memory) {
```
Comment (EvolveArt) [inline solidity/src/libraries/DataParser.sol:11]: probably we should write the specs in comments so it's clearer

File: solidity/src/libraries/MerkleTree.sol [hunk 3]
```diff
@@ -40,26 +41,24 @@ library MerkleTree {
...
         unchecked {
             bytes32 currentDigest = MerkleTree.leafHash(leafData);
-
-            uint8 proofSize = UnsafeCalldataBytesLib.toUint8(encodedProof, proofOffset);
-            proofOffset += 1;
-
-            for (uint256 i = 0; i < proofSize; i++) {
+            uint256 proofOffset = 0;
+            uint16 proofSizeArray = UnsafeCalldataBytesLib.toUint16(encodedProof, proofOffset);
+            proofOffset += 2;
```
Comment (EvolveArt) [inline solidity/src/libraries/MerkleTree.sol:52]: why this change ?

File: solidity/src/libraries/MerkleTree.sol [hunk 4]
```diff
@@ -40,26 +41,24 @@ library MerkleTree {
...
+            for (uint256 i = 0; i < proofSizeArray; i++) {
```
Comment (EvolveArt) [inline solidity/src/libraries/MerkleTree.sol:53]: proofSize was a better name no?

File: solidity/src/libraries/MerkleTree.sol [hunk 5]
```diff
@@ -40,26 +41,24 @@ library MerkleTree {
...
                 bytes32 siblingDigest = bytes32(UnsafeCalldataBytesLib.toBytes32(encodedProof, proofOffset));
-                proofOffset += 20;
+                proofOffset += 32; // TO CHECK
```
Comment (EvolveArt) [inline solidity/src/libraries/MerkleTree.sol:55]: what's to check? We can write tests for this once we have hyperlane validator working right ? let's create an issue so we don't forget

File: solidity/src/libraries/MerkleTree.sol [hunk 6]
```diff
@@ -3,6 +3,7 @@
 pragma solidity ^0.8.0;
 
 import "./UnsafeCalldataBytesLib.sol";
+import "forge-std/console2.sol";
```
Comment (EvolveArt) [inline solidity/src/libraries/MerkleTree.sol]: can be removed

File: solidity/src/Hyperlane.sol [hunk 7]
```diff
@@ -87,18 +82,20 @@ contract Hyperlane is IHyperlane {
             index += 32;
             hyMsg.signatures[i].s = encodedHyMsg.toBytes32(index);
             index += 32;
-            hyMsg.signatures[i].v = encodedHyMsg.toUint8(index) + 27;
+            hyMsg.signatures[i].v = encodedHyMsg.toUint8(index);
```
Comment (EvolveArt) [inline solidity/src/Hyperlane.sol:85]: curious why the +27 is not necessary tho

File: solidity/src/PragmaDecoder.sol [hunk 8]
```diff
@@ -128,56 +140,64 @@ contract PragmaDecoder {
...
             }
-
-            numUpdates = UnsafeCalldataBytesLib.toUint8(encoded, offset);
-            offset += 1;
         }
     }
 
+    function _isProofValid(bytes calldata encodedProof, uint256 offset, bytes32 root, bytes calldata leafData)
+        internal
+        virtual
```
Comment (EvolveArt) [inline solidity/src/PragmaDecoder.sol:152]: why virtual

Comment (EvolveArt) [reply solidity/src/PragmaDecoder.sol:152]: oh ok so it can be overriden for tests ? Worth adding in comment

File: solidity/src/PragmaDecoder.sol [hunk 9]
```diff
@@ -190,26 +210,57 @@ contract PragmaDecoder {
...
-
         unchecked {
             for (uint256 i = 0; i < numUpdates; i++) {
-                DataFeed memory dataFeed;
+                ParsedData memory parsedData;
                 bytes32 dataId;
-                (offset, dataFeed, dataId,) = extractDataInfoFromUpdate(encoded, offset, checkpointRoot);
-                updateLatestDataInfoIfNecessary(dataId, dataFeed);
+                uint64 publishTime;
+                (offset, parsedData, dataId, publishTime) = extractDataInfoFromUpdate(encoded, offset, checkpointRoot);
```
Comment (EvolveArt) [inline solidity/src/PragmaDecoder.sol]: we should probably rename `dataId` to `feedId` to keep things consistent

File: solidity/test/utils/HyperlaneTestUtils.t.sol [hunk 10]
```diff
@@ -29,47 +25,152 @@ abstract contract HyperlaneTestUtils is Test {
...
+        assertEq(parsedMsg.signatures[1].validatorIndex, 1, "Incorrect validator index for second signature");
+        assertEq(uint256(parsedMsg.signatures[1].r), 3, "Incorrect r for second signature");
+        assertEq(uint256(parsedMsg.signatures[1].s), 4, "Incorrect s for second signature");
+        assertEq(parsedMsg.signatures[1].v, 28, "Incorrect v for second signature");
+
+        assertEq(parsedMsg.nonce, 1234, "Incorrect nonce");
+        assertEq(parsedMsg.timestamp, block.timestamp, "Incorrect timestamp");
+        assertEq(parsedMsg.emitterChainId, 5, "Incorrect emitter chain ID");
+        assertEq(uint256(parsedMsg.emitterAddress), 6, "Incorrect emitter address");
+        assertEq(parsedMsg.payload, bytes("Hello, Hyperlane!"), "Incorrect payload");
```
Comment (EvolveArt) [inline solidity/test/utils/HyperlaneTestUtils.t.sol:167]: let's create an issue to test the edge cases of the parsing, especially with special characters or size close to the maxium etc or a very large number of signatures or an invalid timestamp

File: solidity/test/utils/PragmaTestUtils.t.sol [hunk 11]
```diff
@@ -14,7 +14,7 @@ abstract contract PragmaTestUtils is Test, RandTestUtils, HyperlaneTestUtils {
     bytes32 constant SOURCE_EMITTER_ADDRESS = 0x03dA250675D8c2BB7cef7E1b7FDFe17aA4D5752Ed82A9333e4F9a12b22E521aa;
 
     uint256 constant SINGLE_UPDATE_FEE_IN_WEI = 1;
-    uint256 constant VALID_TIME_PERIOD_IN_SECONDS = 60;
+    uint256 constant VALID_timePeriod_IN_SECONDS = 60;
```
Comment (EvolveArt) [inline solidity/test/utils/PragmaTestUtils.t.sol]: revert this change

File: solidity/test/utils/PragmaTestUtils.t.sol [hunk 12]
```diff
@@ -82,7 +82,7 @@ abstract contract PragmaTestUtils is Test, RandTestUtils, HyperlaneTestUtils {
         bytes memory hyperlanePayload = abi.encodePacked(rootDigest);
 
         bytes memory updateData = generateUpdateData(
-            0, config.source_chain_id, config.source_emitter_address, hyperlanePayload, config.numSigners
+            1234, 0, config.source_chain_id, config.source_emitter_address, hyperlanePayload, config.numSigners
```
Comment (EvolveArt) [inline solidity/test/utils/PragmaTestUtils.t.sol]: declare 1234 as a constant like `MOCK_TIMESTAMP_VALUE`

File: solidity/test/DataParser.t.sol [hunk 13]
```diff
@@ -6,102 +6,130 @@ import "../src/libraries/DataParser.sol";
 
 contract DataParserTest is Test {
     function testParseSpotMedianEntry() public pure {
+        bytes32 feedId = bytes32(
+            abi.encodePacked(
+                uint16(0),
+                ///CRYPTO
```
Comment (EvolveArt) [inline solidity/test/DataParser.t.sol]: comment on wrong place

File: solidity/test/DataParser.t.sol [hunk 14]
```diff
@@ -110,57 +138,78 @@ contract DataParserTest is Test {
...
+            uint16(3),
+            uint8(8),
+            uint256(35000 ether), // markPrice
+            uint256(0.001 ether), // fundingRate
+            uint256(1000 ether), // openInterest
+            uint256(500 ether) // volume
         );
-
-        vm.expectRevert("Unknown data type");
+        vm.expectRevert();
```
Comment (EvolveArt) [inline solidity/test/DataParser.t.sol]: don't we have a precise error to test here?

**Review by EvolveArt — COMMENTED** [review]
first round of review

File: solidity/src/libraries/MerkleTree.sol (hunk 5 above)
Comment (JordyRo1) [reply solidity/src/libraries/MerkleTree.sol:55]: Initially is what set to 20 bytes, but from my understanding, unless if we use a specific proof generation process, the proof length should be 32 bytes, that's why I added TO CHECK to be sure. I will create the issue

File: solidity/src/Hyperlane.sol (hunk 7 above)
Comment (JordyRo1) [reply solidity/src/Hyperlane.sol:85]: Usually v is not expressed as y-parity(bool), but as signature scheme(uint8). if y-parity is false, the equivalent signature scheme is often 27 and if the y-parity is true, it's 28. In our case, since it's already displayed as signature scheme, no need to add 27.

File: solidity/src/PragmaDecoder.sol (hunk 8 above)
Comment (JordyRo1) [reply solidity/src/PragmaDecoder.sol:152]: exactly !

File: solidity/src/libraries/MerkleTree.sol (hunk 3 above)
Comment (JordyRo1) [reply solidity/src/libraries/MerkleTree.sol:52]: I made this change because in the previous version, we mixed up the idea of bytes length and array length. The proof is an array of bytes32 stored as bytes, while proofSize is the number of bytes in total in the proof. This mismatch cause the code to fail here. I add the parameter proofSizeArray, which indicates the length of the proof array to avoid this.

File: solidity/src/libraries/MerkleTree.sol (hunk 4 above)
Comment (JordyRo1) [reply solidity/src/libraries/MerkleTree.sol:53]: See above

**Review by EvolveArt — APPROVED** [review]
